### PR TITLE
New advanced client

### PR DESCRIPTION
**Pull request by aalda**

### CODE REVIEW

File: gossip/monitor/monitor.go [hunk 1]
```diff
@@ -68,12 +70,22 @@ func NewMonitor(conf Config) (*Monitor, error) {
 	// Metrics
 	metrics.QedMonitorInstancesCount.Inc()
 
+	// QED client
+	transport := http.DefaultTransport.(*http.Transport)
```
Comment (iknite) [inline gossip/monitor/monitor.go]: this :smile: is what I was meaning in client/options.go:53 comment

**Comment by iknite** [comment]
Also NewHTTPClientFromConfig seems overkill since it's only used twice

```
tests/riot.go
279:	client, err := client.NewHTTPClientFromConfig(cConf)

cmd/client.go
69:		client, err := client.NewHTTPClientFromConfig(clientCtx.config)

```

File: client/config.go [hunk 2]
```diff
@@ -16,42 +16,139 @@
...
-	Leader string
+	// HealthCheckTimeout is the timeout in seconds the healthcheck waits for a response
+	// from a QED server.
+	HealthCheckTimeout time.Duration
 }
 
+// DefaultConfig creates a Config structures with default values.
 func DefaultConfig() *Config {
 	return &Config{
 		Endpoints:               []string{"127.0.0.1:8800"},
```
[the remark this answers is not in the document]
Comment (aalda) [reply client/config.go]: The client constructor also uses the default values. Note that, now, you can construct the client with a conf or with options.

File: client/options.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		}
+
+		defaultTransport := http.DefaultTransport.(*http.Transport)
+		options = append(options, SetHttpClient(&http.Client{
+			Timeout: conf.Timeout,
+			Transport: &http.Transport{
+				Dial: (&net.Dialer{
+					Timeout: conf.DialTimeout,
+				}).Dial,
+				Proxy:                 defaultTransport.Proxy,
```
[the remark this answers is not in the document]
Comment (aalda) [reply client/options.go]: I'm going to get rid of most of these transport options in a future refactoring. Anyway, thanks for the tip...

File: client/retrier.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/bbva/qed/log"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (aalda) [reply client/retrier.go]: Yep, it should be a const.

**Comment by aalda** [comment]
> Also NewHTTPClientFromConfig seems overkill since it's only used twice
> 
> ```
> tests/riot.go
> 279:	client, err := client.NewHTTPClientFromConfig(cConf)
> 
> cmd/client.go
> 69:		client, err := client.NewHTTPClientFromConfig(clientCtx.config)
> ```

The client is also intended to be used outside QED project. I prefer to make the client's construction more flexible.

File: client/topology.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+func (t *topology) Endpoints() []*endpoint {
+	t.Lock()
+	defer t.Unlock()
+	return t.endpoints
+}
+
+// NextReadendpoint returns the next available endpoint to query
+// in a round-robin manner, or ErrNoEndpoint
+func (t *topology) NextReadEndpoint(pref ReadPref) (*endpoint, error) {
```
Comment (gdiazlo) [inline client/topology.go]: This should probably be simplified on a future revision of the client.

**Pull request merged**